core/server/common/src/main/java/alluxio/master/MasterRegistry.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master;
+
+import alluxio.Constants;
+import alluxio.resource.LockResource;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Class for registering individual masters that make up the Alluxio master service. The intended
+ * use of this class is for the Alluxio master service to register the individual masters using an
+ * instance of this class. The registry is passed as an argument to constructors of individual
+ * masters who are expected to add themselves to the registry.
+ *
+ * The reason for using an instance as opposed to a static class is to enable dependency
+ * injection in tests.
+ */
+@ThreadSafe
+public final class MasterRegistry {
+  /**
+   * Used for enforcing the order in which masters are iterated over. If there is a known
+   * dependency between the masters, then the comparator respects the dependency. Otherwise, it
+   * uses alphabetical ordering.
+   */
+  private static final Comparator<Master> COMPARATOR = new Comparator<Master>() {
+    @Override
+    public int compare(Master left, Master right) {
```

Comment:
this isn't transitive so it won't sort the master's correctly, we should do a DFS from every master and add the masters in post-order.